marimo/_plugins/ui/_impl/table.py

```diff
@@ -0,0 +1,153 @@
+# Copyright 2023 Marimo. All rights reserved.
+from __future__ import annotations
+
+from typing import (
+    TYPE_CHECKING,
+    Callable,
+    Final,
+    List,
+    Literal,
+    Optional,
+    Sequence,
+    Union,
+    cast,
+)
+
+from marimo import _loggers
+from marimo._output.mime import MIME
+from marimo._output.rich_help import mddoc
+from marimo._plugins.ui._core.ui_element import UIElement
+
+LOGGER = _loggers.marimo_logger()
+
+Numeric = Union[int, float]
+
+if TYPE_CHECKING:
+    import pandas as pd
+
+TableData = Union[
+    list[dict[str, str]],
+    Sequence[str | int | float | bool | MIME | None],
+    Sequence[dict[str, str | int | float | bool | MIME | None]],
+    "pd.DataFrame",
+]
+
+
+@mddoc
+class table(UIElement[List[str], List[object]]):
+    """
+    A table component.
+
+    **Example.**
+
+    ```python
+    table = mo.ui.table(
+      data=[
+        {'first_name': 'Michael', 'last_name': 'Scott'},
+        {'first_name': 'Dwight', 'last_name': 'Schrute'}
+      ],
+      label='Users'
+    )
+    ```
+
+    ```python
+    # df is a Pandas dataframe
+    table = mo.ui.table(
+        data=df.to_dict('records'),
+        # use pagination when your table has many rows
+        pagination=True,
+        label='Dataset'
+    )
+    ```
+
+    **Attributes.**
+
+    - `value`: the selected values, or `None` if no selection.
+    - `data`: the table data
+
+    **Initialization Args.**
+
+    - `data`:  a list of values representing a column, or a list of dicts
+        where each dict represents a row in the table
+        (mapping column names to values). values can be
+        primitives (`str`, `int`, `float`, `bool`, or `None`)
+        or Marimo elements: e.g.
+        `mo.ui.button(...)`, `mo.md(...)`, `mo.as_html(...)`, etc.
+    - `pagination`: whether to paginate; if `False`, all rows will be shown
+      defaults to `True` when above 10 rows, `False` otherwise
+    - `selection`: 'single' or 'multi' to enable row selection, or `None` to
+        disable
+    - `label`: text label for the element
+    - `on_change`: optional callback to run when this element's value changes
+    """
+
+    _name: Final[str] = "marimo-table"
+
+    def __init__(
+        self,
+        data: TableData,
```

Comment:
I know it's kind of ugly, but can you inline this type? That makes the docs / autocomplete more useful to the user.

Reply:
I think that might also make the type error in the docs CI go away. Otherwise you can replace the `|` with `Union`

Reply:
switched to union, and inline here (where its used in the docs), but kept everywhere else DRYed up